virtualizarr/manifests/store.py

```diff
@@ -387,6 +388,39 @@ def to_virtual_dataset(
             decode_times=decode_times,
         )
 
+    def _find_matching_store(self, request_key: str) -> StoreRequest:
+        """
+        Find the matching store based on the store keys and the beginning of the URI strings,
+        to fetch data from the appropriately configured ObjectStore.
+
+        Parameters:
+        -----------
+        stores : StoreDict
+            A dictionary with URI prefixes for different stores as keys
+        request_key : str
+            A string to match against the dictionary keys
+
+        Returns:
+        --------
+        StoreRequest
+        """
+
+        # Check each key to see if it's a prefix of the uri_string
+        parsed_request_key = urlparse(request_key)
+        for prefix in self._stores.keys():
+            if request_key.startswith(prefix):
+                # Return an existing configured store and parsed request path
+                return StoreRequest(
+                    store=self._stores[prefix], key=parsed_request_key.path
+                )
+        # Use anonymous default store if not in pre-configured stores
+        prefix, store = _default_object_store(request_key)
+        # Add to stores for future use
+        self._stores[prefix] = store
+        self._stores = _sort_stores_by_prefix_length(self._stores)
```

Comment:
Can we just quickly think through the implications of doing this at GET time rather than at `ManifestStore` creation time? It's a little weird that we create a `ManifestStore` which then later edits itself.

Doing it at GET time means if the chunk location URL is nonsense, an error won't be thrown until we actually attempt to read data, which is worse than failing early.

But doing it at `ManifestStore` creation time (which should be possible because the contents of the `ManifestStore` should be immutable once created) means we have to create obstore stores even for URLs that we never intend to read data from. This will happen, for example when converting kerchunk/DMR++ references to Icechunk. I suppose these credentials and so on won't be _tested_ until GET time regardless...

Reply:
I believe Icechunk does something more like the latter. 

The general problem is that overall to read data we need to provide the chunk locations (in the `ChunkManifests`), the store information, and potentially credentials too.

According to @paraseba in icechunk :
> you "configure" your virtual chunk containers in the repo ... with all data but credentials
>
> then you pass the credentials as you open the repo, only when you want to read data

If we did something similar I think that would mean passing (or creating default) obstore stores during `ManifestStore.__init__`, but not requiring credentials at that point. Then later if `.get()` is called then the credentials might be required. Can we do a pattern like that with obstore @kylebarron ?

Reply:
Obstore stores are immutable; you can't change the behavior of a store after construction. Though maybe you could write a custom credential provider, which would be called the first time obstore makes a request.

Reply:
@TomNicholas are there other runtime configuration options that you foresee in VirtualiZarr's future that could impact how we design the configuration of storage instances?

Reply:
I think the only two scenarios that VirtualiZarr has to handle are:
1. you only want to use the `ManifestStore` to represent chunk references in-memory so you don't need credentials to the actual chunk locations,
2. you want to actually use the `ManifestStore` to read chunk data so you do need credentials.

We could have credentials specified or not specified per virtual chunk container like IC supports, but that seems like more of an edge case, the important thing is to be able to avoid specifying credentials at all at store construction time to support case (1).